src/scalar.rs

```diff
@@ -0,0 +1,137 @@
+//! Provides [`Scalar`] and related types.
+//!
+//! In elliptic curve cryptography scalars are non-point values that can be used to multiply
+//! points. The most common type of scalars are private keys. However not all scalars are private
+//! keys. They can even be public *values*. To make handling them safer and easier this module
+//! provides the `Scalar` type and related.
+
+use core::fmt;
+
+/// Positive 256-bit integer guaranteed to be less than the secp256k1 curve order.
+///
+/// The difference between `PrivateKey` and `Scalar` is that `Scalar` doesn't guarantee being
+/// securely usable as a private key.
+///
+/// **Warning: the operations on this type are NOT constant time!**
+/// Using this with secret values is not advised.
+// Internal represenation is big endian to match what `libsecp256k1` uses.
+// Also easier to implement comparison.
+// Debug impl omitted for now, the bytes may be secret
+#[allow(missing_debug_implementations)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Scalar([u8; 32]);
+
+const MAX_RAW: [u8; 32] = [
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE,
+    0xBA, 0xAE, 0xDC, 0xE6, 0xAF, 0x48, 0xA0, 0x3B, 0xBF, 0xD2, 0x5E, 0x8C, 0xD0, 0x36, 0x41, 0x40
+];
+
+impl Scalar {
+    /// Scalar representing `0`
+    pub const ZERO: Scalar = Scalar([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
+    /// Scalar representing `1`
+    pub const ONE: Scalar = Scalar([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]);
+    /// Maximum valid value: `curve_order - 1`
+    pub const MAX: Scalar = Scalar(MAX_RAW);
+
+    /// Generates a random scalar
+    #[cfg(any(test, feature = "rand-std"))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand-std")))]
+    pub fn random() -> Self {
+        Self::random_custom(rand::thread_rng())
+    }
+
+    /// Generates a random scalar using supplied RNG
+    #[cfg(any(test, feature = "rand"))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+    pub fn random_custom<R: rand::Rng>(mut rng: R) -> Self {
+        let mut bytes = [0u8; 32];
+        loop {
+            rng.fill_bytes(&mut bytes);
+            // unlikely to go past MAX
+            if let Ok(scalar) = Scalar::from_be_bytes(bytes) {
+                break scalar;
+            }
+        }
+    }
+
+    /// Tries to deserialize from big endian bytes
+    ///
+    /// **Security warning:** this function is not constant time!
```

Comment:
If there is interest here, you can also add my implementation that is constant time for non-zero values. 

https://github.com/ElementsProject/rust-secp256k1-zkp/pull/16/commits/53d1947d91731ba9b8371900b5a1fd26ee064c0d

I think it makes sense to use the  `_var` suffix naming from upstream to signify variable time.

Reply:
Cool trick! I would like to make it constant time and in pure Rust in the future because it'll also be needed for private key checking without the C library. Looking at the C code, there doesn't seem to be anything C-specific that would make it constant time, so I think it should be doable.

Reply:
Should we rename things to `_var` before cutting a release?

Reply:
I would rather use @sanket1729 trick to not cause churn if we don't want non-constant-time functions without the suffix.
(Also secp256k1 can do the exact check we need but I guess the function is private.)